lib/token/symmetric.go

```diff
@@ -0,0 +1,194 @@
+package token
+
+import (
+	"context"
+	"crypto/aes"
+	"crypto/cipher"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+)
+
+type SymmetricEncoder struct {
+	rng *rand.Rand
+
+	key    []byte
+	cipher cipher.AEAD
+}
+
+type SymmetricSetter func(*SymmetricEncoder) error
+
+// UseSymmetricKey uses a key supplied as an array of bytes.
+//
+// The key can be any array of bytes long enough for the block cipher
+// selected. Use GenerateSymmetricKey to create one.
+func UseSymmetricKey(key []byte) SymmetricSetter {
+	return func(be *SymmetricEncoder) error {
+		be.key = key
+		return nil
+	}
+}
+
+// GenerateSymmetricKey generates a new symmetric key.
+//
+// size is the size of the key to generate in bits. If 0, defaults to 256.
+// The only valid values are those accepted by the underlying AES cipher:
+// 128, 192 or 256.
+func GenerateSymmetricKey(rng *rand.Rand, size int) ([]byte, error) {
+	if size == 0 {
+		size = 256
+	}
+	if size != 128 && size != 192 && size != 256 {
+		return nil, fmt.Errorf("key size is invalid")
+	}
+	key := make([]byte, size/8)
+
+	n, err := rng.Read(key)
+	if err != nil {
+		return nil, err
+	}
+	if n != int(size/8) {
+		return nil, fmt.Errorf("PRNG could not provide %d bytes of key", size)
+	}
+
+	return key, nil
+}
+
+// Creates a new random key and stores it in settings, or return error.
+//
+// size represents the key size in bits. If size is 0, uses a default key size
+// of 256 bits.
+func WithGeneratedSymmetricKey(size int) SymmetricSetter {
```

Comment:
I feel like this one and the other factories should be "WithGenerateSymmetricKeyFromSize"

Reply:
Orthogonal to this PR? Also, that's pretty much the only parameter to generate a symmetric key? In hindsight, this should probably have taken no parameter by default.